File: Code/indifferentcooperator_control.jl
using LightGraphs
using Random
using PyPlot
using StatsBase
using Parameters

function abpath()
    replace(@__DIR__, "Code" => "")
end

function control_indifferentcooperators(numindv, transitionprob, initialprop, time)
    data=zeros(time)
    data2=zeros(time)
    data[1]=numindv*(1-initialprop)
    data2[1]=numindv*initialprop
    for t in 1:time-1
        data[t+1]=data[t]*(1-(transitionprob/numindv))
        data2[t+1]=numindv-data[t+1]
    end
    return hcat(data, data2)
end

let 
    data = control_indifferentcooperators(100, 0.5, 0.05,200)
    test = figure()
    plot(1.0:1.0:200.0,data[:,1], color="blue", label="Indifferent")
    plot(1.0:1.0:200.0,data[:,2], color="orange", label="Cooperative")
    xlabel("Time", fontsize=15)
    ylabel("# of Individuals", fontsize=15)
    legend(fontsize=12)
    # return test
    savefig(joinpath(abpath(), "Output/Images/deterministic_individuals_figure.png"))
end

function resource_regen(δi, δc, Ni, Nc)
    return δi*Ni + δc*Nc
end

function find_time(ind_coop_data, indifferentequalprop)
    numindv = ind_coop_data[1,1]+ind_coop_data[1,2]
    for i in 1:length(ind_coop_data[:,1])
        if isapprox(ind_coop_data[i,1], numindv*indifferentequalprop, atol=0.05)
            return i
        else
        end
    end
end

@with_kw mutable struct Res_Use_Par
    δi = 0.6
    δc = 0.2
end

function control_resources(ind_coop_data, res_regen, indifferentequalprop, res_use_par) 
    @unpack δi, δc = res_use_par
    endtime = find_time(ind_coop_data, indifferentequalprop)
    resourceuse_data = zeros(endtime)
    for i in 1:endtime
        resourceuse_data[i] = res_regen-(δi*ind_coop_data[i,1])-(δc*ind_coop_data[i,2])
    end
    return abs(sum(resourceuse_data))
end

function check_control(numindv, transitionprob, initialprop, time, indifferentequalprop, res_use_par)
    @unpack δi, δc = res_use_par
    res_regen = resource_regen(δi, δc, numindv*indifferentequalprop, numindv*(1-indifferentequalprop))
    ind_coop_data=control_indifferentcooperators(numindv, transitionprob, initialprop, time)
    init_res = control_resources(ind_coop_data, res_regen, indifferentequalprop, res_use_par) 
    resources=zeros(time)
    resources[1]=init_res
    for t in 1:time-1
        resources[t+1]=resources[t]+res_regen-(δi*ind_coop_data[t,1])-(δc*ind_coop_data[t,2])
    end
    return resources
end    

let
    data = check_control(100, 0.5, 0.05, 200, 0.5, Res_Use_Par())
    test = figure()
    plot(1.0:1.0:200.0,data)
    xlabel("Time", fontsize=15)
    ylabel("Resources (units)", fontsize=15)
    ylim(0.0,1040)
    # return test
    savefig(joinpath(abpath(), "Output/Images/deterministic_resources_figure.png"))
end


struct Individual
    node_id::Int
    individual_type::String
end

struct Resources
    value::Float64
    replenishment_proportion::Float64
    consumption_rates::Dict{String, Float64}
end

struct ABMModel
    graph::SimpleGraph
    individuals::Dict{Int, Individual}
    resources::Resources
    p_cooperative::Float64
    beta::Float64
    mu::Float64
    T::Int
    rho::Float64
    critical_value::Float64
    triggered::Bool
    trigger_time::Union{Nothing, Int}
    history::Vector{Dict{String, Float64}}
    resource_history::Vector{Float64}
end

function ABMModel(graph::SimpleGraph, p_cooperative::Float64, beta::Float64, mu::Float64, T::Int,
    initial_resource::Float64, replenishment_proportion::Float64, consumption_rates::Dict{String, Float64},
    critical_value::Float64, rho::Float64)
individuals = Dict(i => Individual(i, "Indifferent") for i in vertices(graph))
# Initialize the population
for i in sample(collect(keys(individuals)), Int(p_cooperative * nv(graph)), replace=false)
individuals[i].individual_type = "Cooperative"
end
resources = Resources(initial_resource, replenishment_proportion, consumption_rates)
ABMModel(graph, individuals, resources, p_cooperative, beta, mu, T, rho, critical_value, false, nothing, [], [])
end

function update!(self::Resources, proportions::Dict{String, Float64})
    total_consumption = 0.0
    # Calculate total consumption based on proportions and consumption rates
    for (ind_type, proportion) in proportions
        total_consumption += self.consumption_rates[ind_type] * proportion
    end

    # Replenish and update resource value
    self.value = self.value - total_consumption + self.replenishment_proportion

    # Ensure resource value is not negative
    if self.value < 0
        self.value = 0
    end
end

function step!(model::ABMModel, t::Int)
    # Trigger extra cooperation if resources fall below the critical value
    if !model.triggered && model.resources.value < model.critical_value
        model.triggered = true
        model.trigger_time = t
    end

    for (i, individual) in model.individuals
        random_value = rand()
        neighbors = neighbors(model.graph, i)
        num_cooperative_neighbors = count(n -> model.individuals[n].individual_type == "Cooperative", neighbors)

        # Probability to become cooperative
        prob_cooperative = 1 - (1 - model.beta) ^ num_cooperative_neighbors

        # Add extra cooperation probability if triggered
        if model.triggered
            prob_cooperative += model.rho
        end

        if individual.individual_type == "Indifferent" && random_value < prob_cooperative
            individual.individual_type = "Cooperative"
        elseif individual.individual_type == "Cooperative" && random_value < model.mu
            individual.individual_type = "Indifferent"
        end
    end
end

function record_state!(model::ABMModel)
    state_counts = Dict("Indifferent" => 0, "Cooperative" => 0)
    for individual in values(model.individuals)
        state_counts[individual.individual_type] += 1
    end

    proportions = Dict(key => count / nv(model.graph) for (key, count) in state_counts)
    push!(model.history, proportions)

    # Update resource based on the proportion of individuals
    update!(model.resources, proportions)
    push!(model.resource_history, model.resources.value)
end

function run!(model::ABMModel)
    for t in 1:model.T
        step!(model, t)
        record_state!(model)
    end
end

# Simulation Parameters
num_individuals = 100
p_cooperative = 0.05
beta = 0.5
mu = 0.2
T = 200

# Resource Parameters
initial_resource = 10.0
replenishment_proportion = 0.46
consumption_rates = Dict("Indifferent" => 1.0, "Cooperative" => 0.1)

# Critical resource parameters
critical_value = 5.0
rho = 0.1

# Create a random graph
graph = SimpleGraph(num_individuals)
for _ in 1:(num_individuals * 0.02 * num_individuals)
    u, v = rand(1:num_individuals), rand(1:num_individuals)
    if u != v
        add_edge!(graph, u, v)
    end
end

# Initialize the model
model = ABMModel(graph, p_cooperative, beta, mu, T, initial_resource, replenishment_proportion, consumption_rates, critical_value, rho)

# Run the simulation
run!(model)

# Extract history
history = model.history
resource_history = model.resource_history

timesteps = 1:T
indifferent_proportions = [h["Indifferent"] for h in history]
cooperative_proportions = [h["Cooperative"] for h in history]

# Plotting
plot(timesteps, indifferent_proportions, label="Indifferent", lw=2, marker=:o)
plot!(timesteps, cooperative_proportions, label="Cooperative", lw=2, marker=:o)
plot!(timesteps, resource_history, label="Resource Value", lw=2, marker=:o, yaxis=:right)

File: Code/indifferentcooperator_control.py
import networkx as nx
import numpy as np
import random
import matplotlib.pyplot as plt
import os

class Individual:
    def __init__(self, node_id, individual_type='Indifferent'):
        self.node_id = node_id
        self.type = individual_type  # Indifferent or Cooperative


class Resources:
    def __init__(self, initial_value, abs_replenishment, consumption_rates):
        """
        Parameters:
        - initial_value: The starting amount of the resource.
        - abs_replenishment: Absolute resources replenishment amount
        - consumption_rates: Dictionary defining how much each type of individual consumes.
        """
        self.value = initial_value
        self.abs_replenishment = abs_replenishment
        self.consumption_rates = consumption_rates

    def update(self, numindv_indcoop):
        """
        Update the resource value based on consumption and replenishment.

        Parameters:
        - numindv_indcoop: A dictionary with the absolute numbers of individuals in each type (Cooperative, Indifferent).
        """
        total_consumption = 0

        # Calculate total consumption based on the proportion of population and individual consumption rates
        for ind_type, numindv_indcoop in numindv_indcoop.items():
            total_consumption += self.consumption_rates[ind_type] * numindv_indcoop

        # Replenishment is a proportion of the current resource level
        self.value = self.value - total_consumption + self.abs_replenishment

        # Ensure resource value does not go negative
        if self.value < 0:
            self.value = 0


    def get_value(self):
        return self.value


class ABMModel:
    def __init__(self, graph, p_cooperative, beta, mu, T, initial_resource, abs_replenishment, consumption_rates):
        """
        Initializes the model with a given graph and other parameters.

        Parameters:
        - graph: The networkx graph to be used in the simulation.
        - p_cooperative: Initial fraction of cooperative individuals.
        - beta: Probability parameter for cooperation.
        - mu: Constant probability of becoming indifferent. (default is zero - CJGG)
        - T: Number of time steps to simulate.
        - initial_resource: Initial value of the resource pool.
        - abs_replenishment: Absolute resources replenishment amount.
        - consumption_rates: Dictionary of consumption rates for each individual type.
        """
        self.graph = graph
        self.num_individuals = graph.number_of_nodes()
        self.p_cooperative = p_cooperative
        self.beta = beta
        self.mu = mu  # Replaces gamma with mu
        self.T = T
        self.individuals = {i: Individual(i) for i in self.graph.nodes()}
        self.initialize_individuals()
        self.history = []  # Store the absolute number of individuals at each time step
        self.resource_history = []  # Store resource values at each step

        # Initialize the Resources class
        self.resources = Resources(initial_resource, abs_replenishment, consumption_rates)

    def initialize_individuals(self):
        """Initialize individuals, all are Indifferent and a fraction p are Cooperative"""
        for i in self.individuals:
            self.individuals[i].type = 'Indifferent'

        # Set initial cooperative individuals based on fraction p_cooperative
        initial_cooperative = random.sample(list(self.individuals.keys()), int(self.p_cooperative * self.num_individuals))
        for i in initial_cooperative:
            self.individuals[i].type = 'Cooperative'

    def step(self, t):
        """Perform a single step of the simulation."""
        for i in self.individuals:
            random_value = np.random.random()
            neighbors = list(self.graph.neighbors(i))
            num_cooperative_neighbors = sum([1 for n in neighbors if self.individuals[n].type == 'Cooperative'])

            # Probability to become cooperative
            prob_cooperative = 1 - (1 - self.beta) ** num_cooperative_neighbors

            # Apply state changes based on probabilities
            if self.individuals[i].type == 'Indifferent':
                if random_value < prob_cooperative:
                    self.individuals[i].type = 'Cooperative'
            elif self.individuals[i].type == 'Cooperative':
                if random_value < self.mu:  # Use mu instead of gamma
                    self.individuals[i].type = 'Indifferent'

    def run(self):
        """Run the simulation for T steps."""
        for t in range(self.T):
            self.step(t)
            self.record_state()

    def record_state(self):
        """Record the proportion of each type of individual and the resource value."""
        state_counts = {'Indifferent': 0, 'Cooperative': 0}
        for individual in self.individuals.values():
            state_counts[individual.type] += 1

        self.history.append(state_counts)

        # Update the resource pool based on the proportion of individuals
        self.resources.update(state_counts)
        self.resource_history.append(self.resources.get_value())

    def get_history(self):
        """Return the history of absolute numbers of indifferent and cooperators at each time step."""
        return self.history

    def get_resource_history(self):
        """Return the history of resource values."""
        return self.resource_history


# Simulation Parameters
num_individuals = 100  # Number of individuals in the graph
p_cooperative = 0.05  # Initial fraction of cooperative individuals
beta = 0.5  # Probability parameter for cooperation
mu = 0.0  # Constant probability of becoming indifferent - DEFAUL is zero CJGG
T = 300  # Number of time steps

# Resource parameters
initial_resource = 1039  # Initial value of the resource (calculated from average/null curve of changing indifferent/cooperators)
abs_replenishment = 40  #Should be Absolute replenishment (calculated from setting consumption rates and number of individuals with 50:50 - indifferent:cooperators)
consumption_rates = {'Indifferent': 0.6, 'Cooperative': 0.2}  # Consumption rates for each type

# Example: Erdos-Renyi graph
graph = nx.erdos_renyi_graph(num_individuals, 0.013)

# Initialize and run the model
model = ABMModel(graph, p_cooperative, beta, mu, T, initial_resource, abs_replenishment, consumption_rates)

# Run the simulation
model.run()

# Get the history of the simulation
history = model.get_history()
resource_history = model.get_resource_history()

# Convert history into time series data for plotting
timesteps = range(T)
indifferent_abs= [h['Indifferent'] for h in history]
cooperative_abs = [h['Cooperative'] for h in history]

# Create a figure with a grid layout: 2 rows (1 plot on the first row, 2 on the second)
fig = plt.figure(figsize=(20, 15))
grid = fig.add_gridspec(2, 2, height_ratios=[1, 1])  # Define grid spec (1st row, centered; 2nd row with two plots)

# First plot: Initial network (spanning the first row)
ax0 = fig.add_subplot(grid[0, :])  # This spans both columns in the first row
ax0.set_title("Initial network config")
pos = nx.spring_layout(graph)  # Layout for the graph
nx.draw(graph, pos, ax=ax0, node_color='gray', with_labels=True, node_size=300, font_size=10)

# Second plot: Proportion of individuals by type
ax1 = fig.add_subplot(grid[1, 0])  # Left plot on second row
ax1.plot(timesteps, indifferent_abs, label='Indifferent', marker='o')
ax1.plot(timesteps, cooperative_abs, label='Cooperative', marker='o')
ax1.set_xlabel('Time Step')
ax1.set_ylabel('Numbers of Individuals')
ax1.set_title('Numbers of Individuals by Type')
ax1.legend()
ax1.grid(True)

# Third plot: Resource consumption over time
ax2 = fig.add_subplot(grid[1, 1])  # Right plot on second row
ax2.plot(timesteps, resource_history, label='Resource Value', color='blue', marker='o')
ax2.set_xlabel('Time Step')
ax2.set_ylabel('Resource Value')
ax2.set_title('Resource Value Over Time')
ax2.legend()
ax2.grid(True)

# Adjust layout and show the figure
plt.tight_layout()

#save the figure

# TODO: Save the figure to a file which is more structured

# PATH_TO_SAVE= "/home/chrisgg/Postdoc/SantaFe/PoliticalEconomyComplexitySchool/TragedyOfTheCommons/Output/Images/V1/"
# plt.savefig(PATH_TO_SAVE+f'sim_ErdosRenyi'+'.png')
plt.show()